### PR TITLE
fix: #6518 Implemented responsive table header controls

### LINKED ISSUE #6518: [Feature]-[500]:The table header controls need to be responsive if there is space available
## Summary

The table widget header controls need to be responsive (until there is space available) in case the widget is resized.

## Motivation

Better UX

## Additional Context

[LOOM DEMO](https://www.loom.com/share/6cd26eb8ccd3481da0ba69615c1314f8)

## Figma link
https://www.figma.com/file/trkiMc12VVpQBqsm7natiU/Table-Changes-Wed-7-Apr-2021?node-id=8%3A7176

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> Made Table header controls responsive when the table widget is made smaller

Fixes #6518

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Tested locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: 6518-responsive-table-header-controls 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.19 **(0.01)** | 37.67 **(0)** | 35.72 **(0.02)** | 56.43 **(0)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :green_circle: | app/client/src/widgets/TableWidget/component/Constants.ts | 86.36 **(0.88)** | 75 **(1.47)** | 91.67 **(0.37)** | 86.36 **(0.88)**
 :green_circle: | app/client/src/widgets/TableWidget/component/Table.tsx | 62.07 **(0.44)** | 38.89 **(0)** | 38.89 **(0)** | 61.45 **(0.47)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableDataDownload.tsx | 35.82 **(0)** | 13.16 **(-0.73)** | 9.09 **(0)** | 34.92 **(0)**
 :red_circle: | app/client/src/widgets/TableWidget/component/TableFilters.tsx | 77.78 **(-2.22)** | 20 **(-3.81)** | 66.67 **(0)** | 76.47 **(-2.32)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableHeader.tsx | 53.49 **(3.49)** | 57.89 **(4.83)** | 36.36 **(6.36)** | 51.22 **(3.85)**
 :green_circle: | app/client/src/widgets/TableWidget/component/TableStyledWrappers.tsx | 87.78 **(0.28)** | 56.12 **(-0.08)** | 81.36 **(0.66)** | 87.78 **(0.28)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
review are always required from internal team members. please make sure to do that in the future

File: app/client/src/widgets/TableWidget/component/TableDataDownload.tsx [hunk 1]
```diff
@@ -202,7 +209,7 @@ function TableDataDownload(props: TableDataDownloadProps) {
         className="t--table-download-btn"
         selectMenu={selectMenu}
         selected={selected}
-        title="Download"
+        title={tableActionTitle}
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/component/TableDataDownload.tsx:212]: Can directly use `props.isMobileScreenTableWidth ? "" : "Download"`. This operation is too simple to be memoized.

File: app/client/src/widgets/TableWidget/component/Table.tsx [hunk 2]
```diff
@@ -126,6 +126,9 @@ export function Table(props: TableProps) {
       }),
     [columnString],
   );
+  const isMobileScreenTableWidth = React.useMemo(() => props.width <= 576, [
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/TableWidget/component/Table.tsx:129]: Directly use props.width <= 576. no need for memo

**Review by sbalaji1192 — APPROVED** [review]

**Comment by shwetha-ramesh** [comment]
@vibhandikyash Tested this PR. Here are the issues I have observed.

1. As per the figma doc and dp does not match. the pagination control does not move when table is resized to extra small
image

2. The controls and do not move smoothly as and when you drag the widget
https://loom.com/share/710d8f39aabb4ec48bdc94bc0b780f30

Moving to Inprogress

**Comment by shwetha-ramesh** [comment]
@vibhandikyash tested this dp. Header controls are now seen on resize. 
image

File: app/client/src/widgets/TableWidget/component/TableFilters.tsx [hunk 3]
```diff
@@ -90,6 +91,14 @@ function TableFilters(props: TableFilterProps) {
     [props.widgetId],
   );
 
+  const tableActionTitle = React.useMemo(
+    () =>
+      `${props.isMobileScreenTableWidth ? "" : "Filters"}${
+        hasAnyFilters ? ` (${filters.length})` : ""
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/component/TableFilters.tsx:97]: we need to check props.isMobileScreenTableWidth to render `(${filters.length})` as well.

File: app/client/src/widgets/TableWidget/component/TableFilters.tsx [hunk 4]
```diff
@@ -90,6 +91,14 @@ function TableFilters(props: TableFilterProps) {
...
+      }`,
+    [props.isMobileScreenTableWidth],
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/component/TableFilters.tsx:99]: Actually we can remove the useMemo. Computation seems to be simple enough to use useMemo

File: app/client/src/widgets/TableWidget/component/TableHeader.tsx [hunk 5]
```diff
@@ -126,45 +130,56 @@ interface TableHeaderProps {
   isVisibleFilters?: boolean;
   isVisiblePagination?: boolean;
   isVisibleSearch?: boolean;
+  isMobileScreenTableWidth: boolean;
   delimiter: string;
+  width: number;
 }
 
 function TableHeader(props: TableHeaderProps) {
+  const isSmallerTableWidth = props.width < TableWidth["Smaller"];
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/component/TableHeader.tsx:139]: <=

File: app/client/src/widgets/TableWidget/component/TableHeader.tsx [hunk 6]
```diff
@@ -207,14 +225,14 @@ function TableHeader(props: TableHeaderProps) {
...
+            {!props.isMobileScreenTableWidth && "Page "}
             <PageNumberInput
               disabled={props.pageCount === 1}
               pageCount={props.pageCount}
               pageNo={props.pageNo + 1}
               updatePageNo={props.updatePageNo}
-            />{" "}
-            of {props.pageCount}
+            />
+            {!props.isMobileScreenTableWidth && ` of ${props.pageCount}`}
```
Comment (sbalaji1192) [inline app/client/src/widgets/TableWidget/component/TableHeader.tsx:235]: remove empty space before of

**Review by sbalaji1192 — COMMENTED** [review]
Please improve PR title and Write cypress cases to confirm that search, download and filters are not appearing in smaller tables.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Resize_spec.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  it("Table Widget check header controls after resize", function() {
+    cy.get(`${commonlocators.layoutControls} span[name='mobile']`).click();
+    cy.get(`${publish.tableWidget} ${publish.searchInput}`).should("not.exist");
+    cy.get(
+      `${publish.tableWidget} ${widgetsPage.tableFilterPaneToggle}`,
+    ).should("not.exist");
+    cy.get(`${publish.tableWidget} ${publish.downloadBtn}`).should("not.exist");
+  });
+
+  afterEach(() => {
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/DisplayWidgets/Table_Resize_spec.js:30]: remove

**Review by sbalaji1192 — COMMENTED** [review]
Since we have made some changes, can we have this QA'ed once again ?
Thanks for making these changes.

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Tested this PR and it is working as expected

**Comment by techbhavin** [comment]
@sbalaji1192   need your approval here

**Review by sbalaji1192 — APPROVED** [review]

**Pull request merged**